### PR TITLE
GH-36323: [Python] Fix Timestamp scalar repr error on values outside datetime range

### LINKED ISSUE #36323: [Python] Repr of Timestamp scalars error if they are outside range of stdlib datetime.datetime
### Describe the bug, including details regarding any error messages, version, and platform.

Currently, the repr of the TimestampScalar (well, of scalars in general) essentially uses the repr of the Python object. For TimestampScalar, this thus fails if this conversion fails, such as for timestamps that are outside of the range of `datetime.datetime` ([1-9999]):

```
>>> res = pa.scalar("0000-01-01").cast(pa.timestamp("s"))
>>>type(res)
pyarrow.lib.TimestampScalar
>>> repr(res)
...
File ~/miniconda3/envs/dev39/lib/python3.9/site-packages/pyarrow/scalar.pxi:117, in pyarrow.lib.Scalar.__repr__()

File ~/miniconda3/envs/dev39/lib/python3.9/site-packages/pyarrow/scalar.pxi:521, in pyarrow.lib.TimestampScalar.as_py()

File ~/miniconda3/envs/dev39/lib/python3.9/site-packages/pyarrow/scalar.pxi:437, in pyarrow.lib._datetime_from_int()

OverflowError: date value out of range

>>> pc.strftime(res)
<pyarrow.StringScalar: '0000-01-01T00:00:00'>
```

Our generic Scalar `__repr__` implementation:

https://github.com/apache/arrow/blob/11b140a734a516e436adaddaeb35d23f30dcce44/python/pyarrow/scalar.pxi#L117-L120

I think for Timestamps we should just convert to a string repr ourselves (`strftime`, or our own pretty printer as is used for the Array repr)

### Component(s)

Python

### PR DESCRIPTION
**Pull request by aboss123**
### Rationale for this change

https://github.com/apache/arrow/issues/36323

### What changes are included in this PR?

Changed the way repr is handled for TimestampScalar

### Are these changes tested?

I have added a very basic test for this change to see whether it will error or not if outside the range.

### Are there any user-facing changes?

The functionality of TimestampScalar's repr now uses the `strftime` function.

* Closes: #36323

### CODE REVIEW

File: python/pyarrow/scalar.pxi [hunk 1]
```diff
@@ -113,8 +113,12 @@ cdef class Scalar(_Weakrefable):
         else:
             with nogil:
                 check_status(self.wrapped.get().Validate())
-
+
```
Comment (kou) [inline python/pyarrow/scalar.pxi:116]:
```suggestion

```

File: python/pyarrow/scalar.pxi [hunk 2]
```diff
@@ -1127,4 +1131,4 @@ def scalar(value, type=None, *, from_pandas=None, MemoryPool memory_pool=None):
 
     # retrieve the scalar from the first position
     scalar = GetResultValue(array.get().GetScalar(0))
-    return Scalar.wrap(scalar)
+    return Scalar.wrap(scalar)
```
Comment (kou) [inline python/pyarrow/scalar.pxi:1134]: Could you revert this change too?

**Review by raulcd — COMMENTED** [review]
There seems to be some linting issues and the following test is failing:
```
__________________ test_sequence_timestamp_from_int_with_unit __________________

    @pytest.mark.pandas
    def test_sequence_timestamp_from_int_with_unit():
        # TODO(wesm): This test might be rewritten to assert the actual behavior
        # when pandas is not installed
    
        data = [1]
    
        s = pa.timestamp('s')
        ms = pa.timestamp('ms')
        us = pa.timestamp('us')
        ns = pa.timestamp('ns')
    
        arr_s = pa.array(data, type=s)
        assert len(arr_s) == 1
        assert arr_s.type == s
>       assert repr(arr_s[0]) == (
            "<pyarrow.TimestampScalar: datetime.datetime(1970, 1, 1, 0, 0, 1)>"
        )
E       assert "<pyarrow.Tim...01T00:00:01'>" == '<pyarrow.Tim... 1, 0, 0, 1)>'
E         - <pyarrow.TimestampScalar: datetime.datetime(1970, 1, 1, 0, 0, 1)>
E         + <pyarrow.TimestampScalar: '1970-01-01T00:00:01'>
```

**Comment by aboss123** [comment]
> There seems to be some linting issues and the following test is failing:
> 
> ```
> __________________ test_sequence_timestamp_from_int_with_unit __________________
> 
>     @pytest.mark.pandas
>     def test_sequence_timestamp_from_int_with_unit():
>         # TODO(wesm): This test might be rewritten to assert the actual behavior
>         # when pandas is not installed
>     
>         data = [1]
>     
>         s = pa.timestamp('s')
>         ms = pa.timestamp('ms')
>         us = pa.timestamp('us')
>         ns = pa.timestamp('ns')
>     
>         arr_s = pa.array(data, type=s)
>         assert len(arr_s) == 1
>         assert arr_s.type == s
> >       assert repr(arr_s[0]) == (
>             "<pyarrow.TimestampScalar: datetime.datetime(1970, 1, 1, 0, 0, 1)>"
>         )
> E       assert "<pyarrow.Tim...01T00:00:01'>" == '<pyarrow.Tim... 1, 0, 0, 1)>'
> E         - <pyarrow.TimestampScalar: datetime.datetime(1970, 1, 1, 0, 0, 1)>
> E         + <pyarrow.TimestampScalar: '1970-01-01T00:00:01'>
> ```

Regarding this error I have a question. The docs appear to want the repr of the TimestampScalar to show datetime but perhaps it could display in a different format? As the initial issue mentions, values that are put in outside the range of the datetime values could break the repr functionality so would it be wise to change docs to show `strftime` output versus the  datetime one?

**Review by danepitkin — COMMENTED** [review]
You're right! We should update documentation as well to reflect this change.

File: python/pyarrow/scalar.pxi [hunk 3]
```diff
@@ -113,8 +113,12 @@ cdef class Scalar(_Weakrefable):
         else:
             with nogil:
                 check_status(self.wrapped.get().Validate())
-
+ 
     def __repr__(self):
+        if isinstance(self, TimestampScalar):
```
[the remark this answers is not in the document]
Comment (aboss123) [reply python/pyarrow/scalar.pxi:118]: Got it!  I can make that change. How would you specify the time format exactly. It mentions the default [here](https://arrow.apache.org/docs/python/generated/pyarrow.compute.strftime.html) but the tests specify `<pyarrow.TimestampScalar: datetime.datetime(2012, 1, 1, 0, 0, tzinfo=<UTC>)>`. This has timezone info would you like the format to include `%z` & `%Z` modifiers to view this?

[the remark this answers is not in the document]
Comment (danepitkin) [reply python/pyarrow/scalar.pxi:118]: Good catch! Yes, let's include the timezone when supplied. Let's try and follow the ISO 8601 standard https://en.wikipedia.org/wiki/ISO_8601 and use `%z`. Maybe something like:

```
>>> res = pa.scalar("2000-01-01").cast(pa.timestamp("s"))
>>> res_with_tz = pc.assume_timezone(res, "America/New_York")
>>> str(pc.strftime(res_with_tz, format='%Y-%m-%dT%H:%M:%S%z'))
'2000-01-01T00:00:00-0500'
```

**Comment by aboss123** [comment]
Given the comments I've made some changes. Please let me know if there is anything else needed.

**Comment by kou** [comment]
@github-actions crossbow submit -g python

**Comment by aboss123** [comment]
@github-actions crossbow submit python

> ```
> Command '['git', '-C', '/tmp/tmpryb2in98/arrow', 'fetch', 'origin', 'pull/36877/head:main']' returned non-zero exit status 128.
> The Archery job run can be found at: https://github.com/apache/arrow/actions/runs/5708248418
> ```

Is there something that needs to be changed considering this did not exit properly?

**Comment by kou** [comment]
Ah, could you re-submit this PR from a feature branch instead of `main`?
Our CI tool (Crossbow doesn't support `main` branch on fork.)

**Comment by aboss123** [comment]
> Ah, could you re-submit this PR from a feature branch instead of `main`? Our CI tool (Crossbow doesn't support `main` branch on fork.)

I have created a new PR here: https://github.com/apache/arrow/pull/36942

**Comment by kou** [comment]
I close this in favor of #36942.
